src/hotspot/share/utilities/bitMap.inline.hpp

```diff
@@ -243,21 +243,39 @@ BitMap::get_next_one_offset_aligned_right(idx_t beg, idx_t end) const {
   return get_next_bit_impl<find_ones_flip, true>(beg, end);
 }
 
-template <typename BitMapClosureType>
-inline bool BitMap::iterate(BitMapClosureType* cl, idx_t beg, idx_t end) {
+template<typename ReturnType>
+struct BitMap::IterateInvoker {
+  template<typename Function>
+  ReturnType operator()(Function function, idx_t index) const {
+    return function(index);
+  }
+};
+
+template<>
+struct BitMap::IterateInvoker<void> {
+  template<typename Function>
+  bool operator()(Function function, idx_t index) const {
+    function(index);
+    return true;
+  }
+};
+
+template <typename Function>
+inline bool BitMap::iterate(Function function, idx_t beg, idx_t end) const {
   for (idx_t index = beg; true; ++index) {
     index = get_next_one_offset(index, end);
     if (index >= end) {
       return true;
-    } else if (!cl->do_bit(index)) {
+    } else if (!IterateInvoker<decltype(function(index))>()(function, index)) {
       return false;
     }
   }
 }
```

Comment:
I added a local function object so that template clutter wasn't all right in the middle of the function.
Making it a member function instead seemed to just push around the deck chairs.
Added comment, and also fixed a lurking, probably never to be hit, bug that we were stopping early if
ReturnType::operator! returned true, rather than if the converted to bool result was false.

Reply:
Here's an alternative, using member function templates instead of the
partially specialized `IterateInvoker` function object. This doesn't seem
obviously better to me. Other ideas I looked at involved messy SFINAE, some
with trailing return types (a not yet approved for HotSpot C++11 feature).

```
template<typename Function, typename Dispatch>
static bool iterate_invoke_aux(Function function, idx_t index, Dispatch*) {
  return function(index);
}

template<typename Function>
static bool iterate_invoke_aux(Function function, idx_t index, void*) {
  function(index);
  return true;
}

template<typename Function>
static bool iterate_invoke(Function function, idx_t index) {
  using ReturnType = decltype(function(index));
  return iterate_invoke_aux(function, index, static_cast<ReturnType*>(nullptr));
}
```